File: PopulatingNextRightPointersInEachNodeII.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None

class Solution:
    # @param root, a tree node
    # @return nothing
    def connect(self, root):
        # Same as I
        if not root:
            return None
        curlist = [root]
        while curlist:
            nextlist = []
            for i in xrange(1, len(curlist)):
                curlist[i-1].next = curlist[i]
            curlist[-1].next = None
            for node in curlist:
                if node.left:
                    nextlist.append(node.left)
                if node.right:
                    nextlist.append(node.right)
            curlist = nextlist

File: CPP/Second/populating_next_right_pointers_in_each_node_ii.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        vector<TreeLinkNode*> curr, next;
        if (!root)
            return;
        curr.push_back(root);
        while (!curr.empty()) {
            next.clear();
            if (curr.size() == 1) {
                curr[0]->next = nullptr;
            } else {
                for (int i = 1; i < curr.size(); i++) {
                    curr[i-1]->next = curr[i];
                }
                curr.back()->next = nullptr;
            }
            for (auto node: curr) {
                if (node->left)
                    next.push_back(node->left);
                if (node->right)
                    next.push_back(node->right);
            }
            curr = next;
        }
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
